Added supplement versions table

javierggt opened the pull request:
## Description

This PR adds two tables to the supplement:
- versions
- last_updated

They have the same structure. Each is a one-row table with string dtype. It is easy to remove the `last_updated` table if you prefer not to have it.

I thought of the following options:
- saving as meta data for each table. Jean wanted a global version as well, plus there should be a method to retrieve versions, and we already have `utils.get_supplement_table`
- saving in a single table, with one row for each existing table, one column for version and another for last_updated. This is fine. It just meant the user code had to jump through some hoops to get the version.

## Testing

- [x] Passes unit tests on MacOS, linux, Windows (at least one required)
- [x] Functional testing. You can run this:
   ```
   (ska3-flight-2021.2rc4) javierg agasc $ export AGASC_DIR=`pwd`/test
   (ska3-flight-2021.2rc4) javierg agasc $ ipython
   Python 3.8.3 (default, Jul  2 2020, 11:26:31) 
   Type 'copyright', 'credits' or 'license' for more information
   IPython 7.16.1 -- An enhanced Interactive Python. Type '?' for help.
   histogram not found

   In [1]: import logging 
      ...: logging.basicConfig(level='DEBUG') 
      ...: import agasc 
      ...: from agasc.supplement.utils import save_version, get_supplement_table 
      ...: from agasc.supplement.utils import save_version, get_supplement_table 
      ...: filename = 'test/agasc_supplement.h5' 
      ...: save_version(filename, obs=agasc.__version__, mags=agasc.__version__, bad=agasc.__version__) 
      ...:                                                                                                                                                      
   DEBUG:agasc.supplement:Creating agasc supplement table "versions"
   DEBUG:agasc.supplement:Creating agasc supplement table "last_updated"
   DEBUG:agasc.supplement:Adding "obs" to agasc supplement "versions" table
   DEBUG:agasc.supplement:Adding "obs" to agasc supplement "last_updated" table
   DEBUG:agasc.supplement:Adding "mags" to agasc supplement "versions" table
   DEBUG:agasc.supplement:Adding "mags" to agasc supplement "last_updated" table
   DEBUG:agasc.supplement:Adding "bad" to agasc supplement "versions" table
   DEBUG:agasc.supplement:Adding "bad" to agasc supplement "last_updated" table
   DEBUG:agasc.supplement:Adding "supplement" to agasc supplement "versions" table
   DEBUG:agasc.supplement:Adding "supplement" to agasc supplement "last_updated" table
   In [2]: get_supplement_table('versions')                                                                                                                     
   Out[2]: 
   <Table length=1>
            obs                   mags                  bad                supplement     
          bytes32               bytes32               bytes32               bytes32       
   --------------------- --------------------- --------------------- ---------------------
   4.10.3.dev89+g8c21663 4.10.3.dev89+g8c21663 4.10.3.dev89+g8c21663 4.10.3.dev89+g8c21663

   In [3]: get_supplement_table('last_updated')                                                                                                                 
   Out[3]: 
   <Table length=1>
             obs                     mags                    bad                  supplement      
           bytes32                 bytes32                 bytes32                 bytes32        
   ----------------------- ----------------------- ----------------------- -----------------------
   2021-02-11 19:11:48.412 2021-02-11 19:11:48.412 2021-02-11 19:11:48.412 2021-02-11 19:11:48.412
   ```

Fixes #74

taldcroft commented on agasc/supplement/utils.py:
```diff
@@ -73,3 +79,60 @@ def get_supplement_table(name, agasc_dir=None, as_dict=False):
+        versions = get_supplement_table('versions')
+
+    :param filename: pathlib.Path
+    :param kwargs: dict
```
Since this function is always being called (except in testing) with `save_version(filename, <table_name>=agasc.__version__)`, maybe just simplify to `save_version(filename, table_name)`?

javierggt replied:
yes, I thought of that, and I also though of having a default argument. I can remove the argument because it makes it simpler, and one can always add it if you want to use a custom version string.

Are you sure you will not want a custom version string? I changed it according to the suggestion, but one can still undo that commit.

taldcroft commented on agasc/supplement/utils.py:
```diff
@@ -73,3 +79,60 @@ def get_supplement_table(name, agasc_dir=None, as_dict=False):
+    time = CxoTime().iso
+    kwargs.update({'supplement': agasc.__version__})
+    for key, value in kwargs.items():
+        pass
```
Leftover.

javierggt replied:
Fixed

taldcroft commented on agasc/supplement/utils.py:
```diff
@@ -73,3 +79,60 @@ def get_supplement_table(name, agasc_dir=None, as_dict=False):
+    version_dtype = '<U32'
+    filename = Path(filename)
+
+    versions = _load_or_create(filename, 'versions')
```
Can we name this table `agasc_versions`? I got quite confused because I think of `versions` as being versions of the tables, not versions of the package that wrote the tables.

javierggt replied:
I changed the table name. We could also add the table version itself, but this is another case where scope is creeping up. The next obvious question is whether we want to track table versions and guarantee back compatibility with older table versions. If we wanted to do that, then perhaps the supplement would have been implemented differently from the ground up.

taldcroft commented on agasc/supplement/utils.py:
```diff
@@ -73,3 +79,60 @@ def get_supplement_table(name, agasc_dir=None, as_dict=False):
+    versions = _load_or_create(filename, 'versions')
+    last_updated = _load_or_create(filename, 'last_updated')
+
+    time = CxoTime().iso
```
Need to check, but I think you can reduce much of the code in the function code to basically:
```
logger.debug(f'Updating {key} in "versions" and "last_updated" tables')
for key, value in kwargs.items():
   versions[key] = [value]
   last_updated[key] = [time]
```
You don't need the `'unknown'`, nor need to pre-declare as columns with a dtype. The key point is that astropy will add in the column or else replace it as the case may be. And I think that the HDF5 writing step will just overwrite whatever is there.

javierggt replied:
Changed to use `CxoTime.now` and set precision.

> You don't need ... to pre-declare as columns with a dtype.

well, I have gotten upset by astropy changing the types when saving/loading fits files (which might be caused by fits, but still) so I decided to have all table types explicit. Anyway, I changed it according to the suggestion.

taldcroft commented on agasc/supplement/utils.py:
```diff
@@ -43,8 +48,9 @@ def get_supplement_table(name, agasc_dir=None, as_dict=False):
 
-    if name not in ('mags', 'bad', 'obs'):
-        raise ValueError("table name must be one of 'mags', 'bad', or 'obs'")
+    if name not in ('mags', 'bad', 'obs', 'last_updated', 'versions'):
```
At this point it may be worth defining a module `AGASC_SUPPLEMENT_TABLES` constant with the names of all the available tables.

javierggt replied:
Done

taldcroft commented on agasc/tests/test_obs_status.py:
```diff
@@ -434,3 +436,26 @@ def recreate_test_supplement():
                                        dry_run=False)
+
+
+def test_save_version(monkeypatch):
```
Nice with the monkey patching!  I haven't really gone down this path before but it looks good.

javierggt replied:
I think we should do more of that in unit testing so it does not depend on the actual SKA/data, and leave the use of SKA/data for integration testing. I think one can also use monkeypatch to generate the data files we would use (say, for cheta)

taldcroft commented on agasc/scripts/update_obs_status.py:
```diff
@@ -76,6 +78,7 @@ def add_bad_star(bad_star_ids, bad_star_source, suppl_file, dry_run):
     logger.info('and installation.')
     if not dry_run:
         dat.write(str(suppl_file), format='hdf5', path='bad', append=True, overwrite=True)
+        save_version(suppl_file, obs=agasc.__version__)
```
I think this should be `bad` not `obs`.

javierggt replied:
Fixed

taldcroft commented on agasc/supplement/utils.py:
```diff
@@ -73,3 +79,60 @@ def get_supplement_table(name, agasc_dir=None, as_dict=False):
+        versions[key][:] = value
+        last_updated[key][:] = time
+
+    versions.write(str(filename), format='hdf5', path='versions',
```
Astropy will accept a `Path`, no need for `str()`here.

javierggt replied:
Fixed

taldcroft reviewed (commented):
One actual bug (I think), an API suggestion and ideas for simplifying / reducing the code (less is more!). Mostly looks good though.

javierggt commented on agasc/supplement/utils.py:
```diff
@@ -73,3 +79,60 @@ def get_supplement_table(name, agasc_dir=None, as_dict=False):
+
+
+def _load_or_create(filename, table_name):
+    try:
+        table = Table.read(filename, format='hdf5', path=table_name)
+    except OSError as e:
+        logger.debug(f'Creating agasc supplement table "{table_name}"')
+        if not filename.exists() or str(e) == f'Path {table_name} does not exist':
+            table = Table()
+        else:
+            raise
```
Actually, the code you propose is not equivalent. What you propose checks if the file exists, while the current code checks whether the table exists in the file.

I agree that relying on exception messages is fragile, but Table.read does not raise a specific exception. One way could be to not use astropy.Table and instead use h5py or pytables to check for the existence of the table.

taldcroft replied:
Ah right I see. Never mind.

javierggt commented:
I just added changes according to all these comments. I hope this is it, because I rebase the branch in #86 in my local copy, fixed conflicts, and started doing some refactoring.

jeanconn commented:
Sure.  I had been thinking one table would be fine, but two is also fine.

jeanconn commented on agasc/scripts/update_obs_status.py:
```diff
@@ -18,26 +18,32 @@
 
-    if not Path(suppl_file).exists():
+    if not suppl_file.exists():
+        if not create:
```
I'm if we have a create flag, the logger message might not be "warning".  But that's a nit.

javierggt replied:
that's right. I didn't think of that. I just left the message that was there (which I had made inconsistent because it said "creating new file" and then raised an exception)

taldcroft reviewed (approved):
Looks good, thanks!!

taldcroft commented:
@javierggt - I'm leaving to you to merge just to be sure.

javierggt commented:
Hey @taldcroft,

I had changed the call to Table.write as you suggested, but now I get this error which results from passing a pathlib.Path instead of a string:
```
In [1]: from pathlib import Path 
   ...: from astropy.table import Table 
   ...: t = Table([{'a':1, 'b':2}]) 
   ...: filename = Path('whatever.h5') 
   ...: t.write(filename, format='hdf5', path='versions', append=True, overwrite=True) 
   ...:                                                                                                                                                      
---------------------------------------------------------------------------
TypeError                                 Traceback (most recent call last)
<ipython-input-1-1045fcc97cd7> in <module>
      3 t = Table([{'a':1, 'b':2}])
      4 filename = Path('whatever.h5')
----> 5 t.write(filename, format='hdf5', path='versions', append=True, overwrite=True)

~/SAO/miniconda3-shiny/envs/ska3-flight-2021.2rc4/lib/python3.8/site-packages/astropy/table/connect.py in __call__(self, serialize_method, *args, **kwargs)
    125         instance = self._instance
    126         with serialize_method_as(instance, serialize_method):
--> 127             registry.write(instance, *args, **kwargs)

~/SAO/miniconda3-shiny/envs/ska3-flight-2021.2rc4/lib/python3.8/site-packages/astropy/io/registry.py in write(data, format, *args, **kwargs)
    561 
    562     writer = get_writer(format, data.__class__)
--> 563     writer(data, *args, **kwargs)
    564 
    565 

~/SAO/miniconda3-shiny/envs/ska3-flight-2021.2rc4/lib/python3.8/site-packages/astropy/io/misc/hdf5.py in write_table_hdf5(table, output, path, compression, append, overwrite, serialize_meta, **create_dataset_kwargs)
    314     else:
    315 
--> 316         raise TypeError('output should be a string or an h5py File or '
    317                         'Group object')
    318 

TypeError: output should be a string or an h5py File or Group object
```

When you told me, I tried, it worked and I though "ah, well, maybe I got it wrong"... but no.

taldcroft commented:
It looks like a bug in astropy, but if you remove the `format='hdf5'` then it works. Go figure. Or you can put the `str()` back in, sorry for the noise.

taldcroft commented:
I checked and this is still a problem in astropy master. @javierggt - can you file a bug report?
